File: frontend/src/api/members.ts
import { apiFetch } from './client'

export type Member = {
  id: number
  name: string
  member_number: string
  street: string
  city: string
  state: string
  zip_code: string
  status: 'active' | 'suspended'
}

export type MemberCreate = Omit<Member, 'id'>
export type MemberUpdate = Partial<Omit<MemberCreate, 'member_number'>>

export function listMembers(q?: string) {
  const qs = q ? `?q=${encodeURIComponent(q)}` : ''
  return apiFetch<Member[]>(`/api/members${qs}`)
}

export function createMember(payload: MemberCreate) {
  return apiFetch<Member>('/api/members', { method: 'POST', body: JSON.stringify(payload) })
}

export function updateMember(memberNumber: string, payload: MemberUpdate) {
  return apiFetch<Member>(`/api/members/${encodeURIComponent(memberNumber)}`, {
    method: 'PUT',
    body: JSON.stringify(payload),
  })
}

export function deleteMember(memberNumber: string) {
  return apiFetch<void>(`/api/members/${encodeURIComponent(memberNumber)}`, { method: 'DELETE' })
}

File: backend/app/routes/members.py
from __future__ import annotations

from fastapi import APIRouter, HTTPException, status

from ..deps import DbDep, require_roles
from ..models.member import Member
from ..schemas.members import MemberCreate, MemberOut, MemberUpdate


router = APIRouter(prefix="/api/members", tags=["members"])


@router.get("", response_model=list[MemberOut], dependencies=[require_roles("manager", "admin")])
def list_members(db: DbDep, q: str | None = None):
    query = db.query(Member)
    if q:
        like = f"%{q.strip()}%"
        query = query.filter((Member.name.ilike(like)) | (Member.member_number.ilike(like)))
    return query.order_by(Member.name.asc()).all()


@router.post("", response_model=MemberOut, status_code=status.HTTP_201_CREATED, dependencies=[require_roles("manager", "admin")])
def create_member(payload: MemberCreate, db: DbDep):
    exists = db.query(Member).filter(Member.member_number == payload.member_number).one_or_none()
    if exists:
        raise HTTPException(status_code=400, detail="Member number already exists")
    m = Member(**payload.model_dump())
    db.add(m)
    db.commit()
    db.refresh(m)
    return m


@router.put("/{member_number}", response_model=MemberOut, dependencies=[require_roles("manager", "admin")])
def update_member(member_number: str, payload: MemberUpdate, db: DbDep):
    m = db.query(Member).filter(Member.member_number == member_number.strip()).one_or_none()
    if not m:
        raise HTTPException(status_code=404, detail="Member not found")
    for k, v in payload.model_dump(exclude_unset=True).items():
        setattr(m, k, v)
    db.commit()
    db.refresh(m)
    return m


@router.delete("/{member_number}", status_code=status.HTTP_204_NO_CONTENT, dependencies=[require_roles("manager", "admin")])
def delete_member(member_number: str, db: DbDep):
    m = db.query(Member).filter(Member.member_number == member_number.strip()).one_or_none()
    if not m:
        raise HTTPException(status_code=404, detail="Member not found")
    db.delete(m)
    db.commit()
    return None

